### ISSUE TITLE
Reusable tasks

### ISSUE THREAD

**Issue by vito** [concourse/rfcs#7]
(this is a rough draft/placeholder just to centralize brainstorming before an actual RFC)

ref. concourse/concourse#417
ref. concourse/docker-image-resource#185

There is clearly a need to make tasks just as easy to share, version, and re-use just as resources are.

What would this look like?

Today you can approximate this by storing task `.yml`s in a Git repo, and then pulling them in as a resource.

One big hole is configuration. You can only really specify `params`, which for tasks are environment variables. Lots of resources use `params` for arbitrary, sometimes nested config. Should we extend tasks to have the same ability? It would actually be pretty easy to just pass it to the script on `stdin`, just like we do for resources. (Would we then rename today's `params` to `env` or something to be more clear?)

**Comment by jchesterpivotal** [comment]
A few months backed [I sketched some notes for these](https://docs.google.com/document/d/1hszTDI1FpZfMeycDsTk3cXQ-w4DVATXjPPUYhlxBcGw/edit?usp=sharing). It goes a bit further than what's immediately needed -- I assume that streaming inputs are a supported case.

**Comment by vito** [comment]
I wanna pull https://github.com/concourse/concourse/issues/608 into the discussion here too. I think if we had more shareable tasks we'd be more inclined to implement things like the Terraform resource as a task or set of tasks instead of a resource, and then the user is free to use whatever storage backend the like for persisting the task's side effects.

I recognise that this goes against the idea of tasks being pure and side-effect-free, but I think the more important aspect has always been that they're *idempotent*, and I'm interested in seeing if this change in mindset leads to a more expressive reconfiguration of existing ideas.

I've mocked up a pipeline for this here: https://gist.github.com/vito/3f0ed09648bf117d95a3b7daa99f61c2

Here are some decisions I made with this mockup:

* Tasks are defined at the top-level, and are referenced by name when used in jobs. Now there's an actual use for the name of the task step! That feels nice an intuitive - task names have been completely cosmetic for a while, unlike their `get`/`put` counterparts.
* Top-level tasks are provided by a resource, similar to resource types. This supports storing them in `git` repos pretty well. This pattern is already followed by [cf-deployment-concourse-tasks](https://github.com/cloudfoundry/cf-deployment-concourse-tasks) for example. I like storing tasks in `git` repos as it makes versioning and inspection very easy (browse around on GitHub rather than reverse-engineering a `Dockerfile`).
* Each top-level task configures the path to the task config file within its resource. Theoretically this could be inferred from the task name and we could have a convention like `tasks/(name).yml` - having to type it out feels a little clunky compared to resources. Feels kinda magical tho.
* I snuck in a new feature on `get` steps: `optional: true`. This would be necessary for the initial deploy where there's no state yet. This feature would have obviated the need for https://github.com/concourse/s3-resource/issues/21 which was suggested precisely for this very use case. (Also, tasks already support optional inputs so these two features would work nicely together.)

One thing I decided against with this mockup was having tasks provided by Docker images (a-la resource types). In analyzing my current use case and `cf-deployment-concourse-tasks`, I noticed that there tends to be more than one task using a given image. For example, my pipeline already has a case for both `terraform apply` and `terraform output`. Forcing those to be entirely separate images feels wasteful.

There'd also be a bit of a circle here because task configs already exist as a top-level task entrypoint, and they can already point to their own image. So it felt more natural and flexible to just start from a task config rather than a Docker image.

**Comment by jchesterpivotal** [comment]
A nitpick -- if we do allow inferred directories, I'm keen to push on using the full taskdir pattern. So instead of `tasks/task-name.yaml`, it's `tasks/task-name/task.yaml`. The executable by convention is `tasks/task-name/run` (not `task`, because tab completion is nice).

In my experience tasks can become quite sophisticated; having all the logic isolated in the subdirectory is easier to navigate and selectively update.

**Comment by jchesterpivotal** [comment]
A pattern I am trying is to bundle tasks with a resource image, as they are intended to be used together. See [concourse-build-resource](https://github.com/jchesterpivotal/concourse-build-resource) for the example.

**Comment by ari-becker** [comment]
Internally, we're approaching this problem with [Dhall](https://dhall-lang.org/). If the rest of the pipeline is also written in Dhall, then we can write a default "implementation" for the task, pull it in, and pass it configuration by way of function parameters or by overwriting the relevant fields of the default.

More broadly, this may not just be re-usable tasks, but re-usable pipelines - allowing people to write common pipelines for workflows like building, testing, and pushing containers, where the entire pipeline is represented by a function which takes parameters like the image name, registry details, etc.

It would be great if Concourse had better support for Dhall (we're currently using `dhall-to-yaml` to get it to work with Concourse), but maybe that's a bigger discussion.

**Comment by amnk** [comment]
@ari-becker wow. Are there any public resources about that approach? Maybe some articles you wrote? Or at least a gist with an example? :)

**Comment by ari-becker** [comment]
> Are there any public resources about that approach? Maybe some articles you wrote? Or at least a gist with an example? :)

@amnk For Concourse, not yet. Right now we're trying to ship our implementation internally, then it needs to be cleaned up / refactored, and then we'll get around to open-sourcing the generic bits and presenting on it. It can't really be summed up in a gist because Dhall ends up forcing you to write types for the entire schema. But you can imagine that if you had a task type:

```dhall
{ task : Text
, config : Optional 
  { platform : Text
  , image_resource : Optional ImageResource
  , inputs : Optional  (List Input)
  , run : Run
  , rootfs_uri : Optional Text
  , outputs : Optional (List Output)
  , caches : Optional (List Cache)
  , params : Optional (List {mapKey : Text , mapValue : Text})
  }
, file : Optional Text
, privileged : Optional Bool
, params : Optional (List {mapKey: Text, mapValue: Text})
, image : Optional Text
, input_mapping : Optional (List {mapKey : Text, mapValue : Text})
, output_mapping : Optional (List {mapKey : Text, mapValue : Text})
}
```

where `ImageResource`, `Input`, `Output`, and `Cache` are defined elsewhere, and you wrote a default task for some computation:

```dhall
\(_params : { config_params : List {mapKey : Text, mapValue : Text}, task_params : List {mapKey : Text, mapValue: Text}}) ->
{ task = "example-computation"
, config = Some 
  { platform = "linux"
  , image_resource = Some <fill_in_the_gap...>
  , inputs = Some [ <input_dir> ]
  , run = <does_the_computation>
  , rootfs_uri = None Text
  , outputs = Some [ <output_dir> ]
  , caches = None (List Cache)
  , params = Some _params.config_params
  }
, file = None Text
, privileged = None Bool
, params = Some _params.task_params
, image = Some "<image_defined_elsewhere>"
, input_mapping = None (List {mapKey : Text, mapValue : Text})
, output_mapping = None (List {mapKey : Text, mapValue : Text})
} : Task
```

Then all the consumer has to do is call :

```dhall
mkDefaultExampleComputationTask { config_params = [ { mapKey = "example_key", mapValue = "example_value" }], task_params =  [ { mapKey = "example_key", mapValue = "example_value" }] } // { image = Some "<other_image>" }
```

and if you call `dhall-to-yaml` on that then you get your generated task as yaml, with the given parameters from `config_params` and `task_params`, and the image field will show `<other_image>` instead of `<image_defined_elsewhere>`.

Now, idiomatic Dhall doesn't really look like that; the types are handled differently, and you can write more than one implementation, e.g. `mkParamlessExampleComputationTask` which inlines empty lists for `params` in both the task and the config.... lots of stuff like that. But hopefully that gives some idea of where to get started.

**Comment by itsdalmo** [comment]
> I recognise that this goes against the idea of tasks being pure and side-effect-free, but I think the more important aspect has always been that they're idempotent, and I'm interested in seeing if this change in mindset leads to a more expressive reconfiguration of existing ideas.

I think this RFC and some help for reusable tasks is necessary, but I think the distinction between a `task` and a `resource` becomes somewhat blurry in this case. Since resources and tasks are both meant to be idempotent, and both can have side-effects, what then is the distinction between a task and a resource?

And what do we mean by idempotent here? Using `terraform` as an example, strictly speaking it is impossible for the task to be idempotent because the state-file is opaque to Concourse, and the true state is not known until terraform refreshes the state?

>One thing I decided against with this mockup was having tasks provided by Docker images (a-la resource types). In analyzing my current use case and cf-deployment-concourse-tasks, I noticed that there tends to be more than one task using a given image. For example, my pipeline already has a case for both terraform apply and terraform output. Forcing those to be entirely separate images feels wasteful.

A simple solution to the above would just be to pass `terraform apply` or `terraform output` as arguments to the task, which would then allow us to reuse the same image for multiple tasks. I think for tasks like `terraform apply` (which has potentially large side-effects) and other tasks where correctness is paramount, it makes the most sense if the task itself is responsible for setting defaults and validating the source configuration and parameters, instead of relying on a `task.yml` which may/may not get out of sync with the code (especially if the code is delivered as part of a docker image, in the case of e.g. compiled languages).

I think this RFC is nice for shaping up tasks as they exist today, but I feel like we are still missing a solution for "serious" tasks as mentioned above. Essentially a resource which implements only `put`, which is today considered an anti-pattern, would be perfect for something like `terraform apply` and friends 🤔

**Comment by ari-becker** [comment]
> @ari-becker wow. Are there any public resources about that approach? Maybe some articles you wrote? Or at least a gist with an example? :)

Just to link to it from this issue: since replying to @amnk , we've open-sourced [dhall-concourse](https://github.com/coralogix/dhall-concourse), which provides a limited solution to this problem, and I imagine is a helpful reference point for the maintainers looking at solutions.

> I think this RFC and some help for reusable tasks is necessary, but I think the distinction between a `task` and a `resource` becomes somewhat blurry in this case. Since resources and tasks are both meant to be idempotent, and both can have side-effects, what then is the distinction between a task and a resource?

A resource defines some/all of three named impure tasks: `check`, `in`/`get`, and `out`/`put`, each having fixed, defined input and output schemas so as to be additionally useful to Concourse. I imagine that you could think of resources as a unique, built-in to Concourse, reusable union of those three named impure tasks. That's not quite the same thing as generic, reusable, ideally-pure tasks.

> ... Terraform ...

Terraform is nearly idempotent when plan files are used - essentially, state can be versioned for the `check` by hashing the output of `terraform plan`, state can be fetched for `in`/`get` by running `terraform plan -out=terraform.plan` and making sure that it matches the hash, and state can be pushed for `out`/`put` by running `terraform apply terraform.plan`. It's certainly not a `put`-only paradigm.

It's still not a great fit for Terraform though, because `terraform apply` cannot (even in theory) be represented according to Concourse's current version paradigm - for a given invocation of `terraform apply`, whether using a plan file or not, and for reasons entirely outside of Terraform's or Concourse's control (i.e. provider availability), the output may represent success, failure, or somewhere in the middle (what Terraform calls "tainted"). In essence, Concourse currently thinks of versions as being entirely deterministic, which doesn't fit Terraform's model because Terraform won't know what the "version" of the infrastructure will be until after `terraform apply` is called, not before. Anyway, Concourse's notions of versioning don't seem so relevant to this RF(RFC).

**Comment by itsdalmo** [comment]
> A resource defines some/all of three named impure tasks: `check`, `in`/`get`, and `out`/`put`, each having fixed, defined input and output schemas so as to be additionally useful to Concourse. I imagine that you could think of resources as a unique, built-in to Concourse, reusable union of those three named impure tasks. That's not quite the same thing as generic, reusable, ideally-pure tasks.

Thanks for explaining, but I don't feel like this makes it any more clear to me. Currently, I could write a task with side effects, it just would not be versioned by Concourse. Likewise, I could write a resource which is not versioned by Concourse by implementing a no-op `check` and `get`, and not emitting any versions from `put`. Both the task and the resource would behave the same way from a user perspective, but Concourse would treat them differently internally?

Anyway, I'm not saying that Concourse does not need tasks. I'm just thinking that the [`info` for resources v2](https://github.com/concourse/rfcs/pull/1/files#diff-82f93c150101c77aca4bf38b2ee5312dR5) would allow us to implement a wider range of resource types, and have Concourse work out internally which keywords map to which operation, and how to use them. So e.g. we could add a `task` artifact to a resource, and Concourse would know that said resource could be used for tasks, and that it would not output any versions for Concourse to track when used as a `task`.

> Terraform is nearly idempotent when plan files are used - essentially, state can be versioned for the check by hashing the output of terraform plan, state can be fetched for in/get by running terraform plan -out=terraform.plan and making sure that it matches the hash, and state can be pushed for out/put by running terraform apply terraform.plan. It's certainly not a put-only paradigm.

I don't think this is correct/would work for the following reasons:
- The output of `terraform plan` and the hash changes on every invocation even if the plan stays the same. Additionally, it is in a binary format and it is documented as having an unstable API and being solely for consumption by `terraform`.
- `terraform plan` does not fetch the state, it does however refresh the current state before producing a plan which describes how to get from `current state > desired state`. As such, using `terraform plan` and `apply` in two separate steps does not really make it any more idempotent since the state lock would be released between `plan` and `apply`, which means that both the terraform state and the "true" state could have changed in between.

**Comment by vito** [comment]
> I think this RFC and some help for reusable tasks is necessary, but I think the distinction between a `task` and a `resource` becomes somewhat blurry in this case. Since resources and tasks are both meant to be idempotent, and both can have side-effects, what then is the distinction between a task and a resource?

Resources are versioned, tasks are not. Tasks can perform side-effects, but they can only produce artifact outputs, and to persist them or version them, you would *then* use a resource to persist those outputs. This is a more explicit separation and lets people decide how to persist things like terraform state without the terraform resource having to implement multiple persistence backends.

I see this as our tactic to eliminate `put`-only resources - especially as we'll soon be removing their ability to even record versions in light of https://github.com/concourse/concourse/issues/3463, as that has led to a lot of user confusion.

> And what do we mean by idempotent here? Using `terraform` as an example, strictly speaking it is impossible for the task to be idempotent because the state-file is opaque to Concourse, and the true state is not known until terraform refreshes the state?

All I mean by idempotent is the task can be re-run with the same inputs and will attempt to do the same thing. Partial successes can be re-tried, and may eventually succeed. It sounds like the conflict here is that even partial successes have to be persisted in some way, though I'm really not sure how to reason about that because you could always fail to persist the state file anyway.

**Comment by vito** [comment]
I forgot to note it here, but my most recent experiment in this area is the [`builder` task](https://github.com/concourse/builder-task).

To be honest, it already works pretty well! It's been interesting seeing the patterns emerge from how it's used. The task image pulled in via a resource in our pipeline and used like this:

https://github.com/concourse/concourse/blob/9d2a80765400d89b0214a396b7532b1245675e39/ci/pipelines/concourse.yml#L93-L105

...but then we also have some task .ymls that pull it in via `image_resource`:

https://github.com/concourse/concourse/blob/master/ci/tasks/build-rc-image.yml#L4-L7

This makes it easy to use `fly execute` to run the task and allows more of its configuration to be pre-defined.

Here are things I learned from this experiment:

* Allow images to configure an `ENTRYPOINT` - then consumers wouldn't need to set `run:` at all in this case, and Terraform could just have an entrypoint that interprets `apply` or `output` arguments.
* Being able to use JSON instead of env for `params:` would go a long way in terms of API flexibility. This would bring tasks up to par with resources.
* Task caches are actually pretty useful, at least in our scenario - it lets us cache intermediate layers between image builds. This was a pleasant surprise.
* `input_mapping` and `output_mapping` work nicely with generic tasks, and being able to list the inputs in the task config helps a lot.

It feels like there aren't many missing pieces here, and that by implementing those first two points we might enable some pretty interesting usage patterns that integrate nicely with a lot of today's existing patterns.

**Comment by ljfranklin** [comment]
Really liking the direction this is heading, specifically the ability to ship tasks as docker images rather than repos with random bash scripts and the ability to pass YAML/JSON rather than trying to pack complex structures into env vars.

Bit of a tangent but I recently made some changes to the [terraform-resource](https://github.com/ljfranklin/terraform-resource) that seems related. Up to this point the terraform-resource only supported storing state in S3-compatible blobstores which was a pain for users that would rather use GCS/Azure/etc. However Terraform recently added support for storing the statefile in a variety of [backends](https://www.terraform.io/docs/backends/) from the CLI itself. There are currently 10+ backends which are supported. Once you configure a backend, when you run `terraform apply` the CLI automatically persists the statefile changes into the configured storage endpoint. I recently updated the resource to use this backend functionality rather than my home-rolled S3 client.

Along with these backends Terraform also now has a concept of [workspaces](https://www.terraform.io/docs/state/workspaces.html). This allows Terraform to store multiple statefiles within a single backend, as well as list, download, delete those statefiles. These changes have nicely filled out the terraform-resource so it can more realistically implement the `get` and `check` operations. It still falls short in fetching versions other than `latest` for each environment, but at least now it doesn't have to rely on `put` producing version numbers ([proposal to remove put-only resource functionality](https://github.com/concourse/concourse/issues/3463)).

While the re-usable tasks are going to be super useful, for the Terraform case specifically I'm still liking the resource as an interface. The user doesn't have to worry about getting the `ensure` blocks just right in order to persist state in the event of a failure and using `backends` allows you to specify your storage configuration via a first-class Terraform feature so you're more in line with the rest of that community. I'm also still having trouble letting go of my "ideal" task mental model:
- `get` on a resource pulls state from the outside world (e.g. the internet)
- `put` on a resource modifies state in the outside world
- tasks are purely functional operations on inputs and outputs, no side-effects
  - also makes tasks easily unit testable if you want to test your test scripts

In practice I've written a ton of tasks that don't conform to this description as it was easier to just write a bunch of `wget`s than fit what I was doing into a resource, hence the scare quotes on "ideal". All that said, none of this disagrees with the re-usable task proposal above, just thinking out loud around what could/should be modeled as a task vs a resource.

**Comment by vito** [comment]
@ljfranklin Thanks, that context really helps! :+1: Using Terraform's persistence backends definitely makes it feel more "resource"-like.

**Comment by hstenzel** [comment]
I like the direction this is going, but I also think there's another angle of analysis. I came to this discussion from https://github.com/concourse/docker-image-resource/issues/249. You see, what I was trying to accomplish is something like a 'factory' pattern.

The Docker image resource seems so close to providing what I needed, which is to do a `docker build` then grab an artifact (**not** the resulting image) and push just the artifact to a repository using another resource. Of course, we can do that today with the Docker image resource but only if we make a round trip to the external repository.

Tasks don't have this problem since inputs and outputs can be defined, but there still seems to be some kinds of resources that may have a `task` capability as well.

Yes, the problem can also be solved by building into the pipeline the ability to generate the image that is then used in a task; but this ends up being unreliable because of the racy and distributed nature of docker image repositories.

Fundamentally, we're trying to put the build logic into the repo in a way that it doesn't care if it's exercised as part of a pipeline or on a developer workstation. For this, it seems like controlling the task with a Dockerfile and outputting a rootfs may be another way of accomplishing a similar goal.

This is very task-y in the sense that it would have the inputs and outputs, but very docker-y in the sense that how the build is performed is controlled by a Dockerfile, args, etc.

Thoughts?

**Comment by vito** [comment]
> The Docker image resource seems so close to providing what I needed, which is to do a `docker build` then grab an artifact (**not** the resulting image) and push just the artifact to a repository using another resource. Of course, we can do that today with the Docker image resource but only if we make a round trip to the external repository.

I'm having a hard time understanding this part. Would you mind attaching a pipeline snippet or something? :thinking:

**Comment by blckct** [comment]
@vito Some projects have a build system in docker where the docker build pulls all the build tools, the dependencies and then spews out the resulting output. Seems they only want the resulting output and not the docker image. Like, run docker build and then get a jar file in the current directory at the end.

**Comment by hstenzel** [comment]
Sure. Let's say I want to build a jar. But I want to put all the steps into a Dockerfile because it gives a good developer experience **and** it works well in CI type environments.

Right now I can:

```
jobs:
- name: build
  - get: src
    trigger: true
  - put: docker-image-with-jar
    params:
      build: src
  - get: docker-image-with-jar
  - put: release-maven-jar
      file: docker-image-with-jar/rootfs/*.jar
      pom_file: docker-image-with-jar/rootfs/pom.xml
```

But in order to do that, I have to first publish the docker image to a repository. Ideally, the `get docker-image-with-jar` and `put docker-image-with-jar` would be replaced by the proposed new reusable task.

**Comment by vito** [comment]
@hstenzel Seems like that is possible now thanks to https://github.com/concourse/builder-task/pull/25:

```
~/workspace/builder-task master*
❯ cat my-asset-builder/Dockerfile 
FROM alpine
RUN mkdir /my-asset && echo 'hello' > /my-asset/some-file

~/workspace/builder-task master*
❯ cat example.yml
---
platform: linux

image_resource:
  type: registry-image
  source:
    repository: concourse/builder-task
    tag: alpine-based

params:
  REPOSITORY: doesnt/matter
  TAG: latest
  CONTEXT: my-asset-builder
  UNPACK_ROOTFS: true

inputs:
- name: my-asset-builder

outputs:
- name: my-asset
  path: image/rootfs/my-asset

run: {path: build}

~/workspace/builder-task master* 14s
❯ fly -t l e -c example.yml -i my-asset-builder=my-asset-builder/ -o my-asset=/tmp/my-asset -p
uploading my-asset-builder done
executing build 22 at http://localhost:8080/builds/22 
initializing
running build
building
Building docker.io/doesnt/matter:latest
Setting up the rootfs... this may take a bit.
[+] Building 2.0s (6/6) FINISHED                                                
 => [internal] load .dockerignore                                          0.1s
 => => transferring context: 2B                                            0.0s
 => [internal] load build definition from Dockerfile                       0.1s
 => => transferring dockerfile: 113B                                       0.0s
 => [internal] load metadata for docker.io/library/alpine:latest           0.8s
 => [1/2] FROM docker.io/library/alpine@sha256:769fddc7cc2f0a1c35abb2f914  0.6s
 => => resolve docker.io/library/alpine@sha256:769fddc7cc2f0a1c35abb2f914  0.0s
 => => sha256:769fddc7cc2f0a1c35abb2f91432e8beecf83916c42 1.64kB / 1.64kB  0.0s
 => => sha256:bf1684a6e3676389ec861c602e97f27b03f14178e5bc3f7 528B / 528B  0.0s
 => => sha256:e7c96db7181be991f19a9fb6975cdbbd73c65f4a268 2.76MB / 2.76MB  0.3s
 => => sha256:055936d3920576da37aa9bc460d70c5f212028bda1c 1.51kB / 1.51kB  0.0s
 => => unpacking docker.io/library/alpine@sha256:769fddc7cc2f0a1c35abb2f9  0.2s
 => [2/2] RUN mkdir /my-asset && echo 'hello' > /my-asset/some-file        0.3s
 => exporting to image                                                     0.2s
 => => exporting layers                                                    0.1s
 => => exporting manifest sha256:7ee95253eebf81cbbbcf40537b8aad266bb1ac9b  0.0s
 => => exporting config sha256:efd4252088147d5c3d46b86bc0a4280734fb9d8c50  0.0s
 => => naming to docker.io/doesnt/matter:latest                            0.0s
Successfully built docker.io/doesnt/matter:latest

saving image

unpacking rootfs
Successfully unpacked rootfs for doesnt/matter:latest to: /scratch/rootfs
succeeded
downloading my-asset done

~/workspace/builder-task master* 27s
❯ ls /tmp/my-asset/
some-file
```

**Comment by hstenzel** [comment]
@vito Very cool, thanks! I'll try it out!

**Comment by jchesterpivotal** [comment]
Another thing that struck me today is that it would be super helpful to allow tasks to emit metadata, comparably to resources. Not very useful _today_, but in a world of widely-shared tasks, could become quite useful.

For example, you can imagine a `run-ginkgo` task that has metadata about the number of passed/failed/errored, plus an array of failed tests. This would greatly simplify the business of extracting test flakiness information. Buildpacks tasks could emit layer metadata.

I see three main problems:
1. Database bloat. But that's a thing no matter what we do.
1. Implementation. How the heck would a task signal the metadata? STDOUT and STDERR are already in use. Options might include writing to a known file location, setting environment variables, or providing a second post-task script that emits metadata.
1. Blurring the line further between task and resource. I worry about this least, but it's a likely effect.

I don't think such a change would be part of a first RFC about redistributable tasks. But I can see it as a followup feature.

**Comment by jchesterpivotal** [comment]
Yes, I think having a consistent pattern would be best.

**Comment by hstenzel** [comment]
I'm really liking the promise of the `builder-task`. In thinking about it though, I wonder if a different formulation would be better.

Consider: aren't we really just talking about curated toolchain images?

In this formulation, the `builder-task` would be the curated `img-toolchain-task`. This is the container image that has a comfortable toolchain to build a wide variety of images using `img`.    

- It would have everything needed to build docker images with `img` (which it already does)
- It would have a smattering of other tools -- think jq, make, busybox, etc (it already has many)
- It would be its own project and would be stably versioned (not too far off)
- It would have a `build` command that wraps `img` (as it already does)
- The `build` command could wrap other concourse integration constructs. Examples might include:
  - Adding build metadata as labels ( `--concourse-add-metadata` )
  - computing reasonable image names due to knowledge ( `--concourse-tag "%b-%s"`for branch and sha)
- Other commands or utilities to make it easier to integrate with Concourse could be added

With this formulation, there would not be any need for a full task resource api, there would be excellent opportunity for reuse, and the build process could still work in a natural way both in CI and non-CI environments. Most importantly, there would be an off-the-shelf solution for Docker (which seems to be the biggest hitter due to how common it is and due to how complex it is to get a builder working.

It would also be easy to imagine other curated task images that would contain just the toolchain for other popular stacks.

In other words, the consistent pattern **is** the task pattern! There's just reuse put in place around it.

Thoughts?

**Comment by jchesterpivotal** [comment]
I've had in the back of my mind that once reusable tasks arrived that libraries of a sort would quickly emerge (even for stuff as simple as `cp`). But I'd be wary of overspecifying what prospective task implementers need to implement to be conformant. We create a narrow doorway, yes, but a narrow doorway needs less paint.

As for build tasks _in particular_, my horse in this race is [Cloud Native Buildpacks](https://buildpacks.io/). I'm aware of at least one POC that uses the `pack` CLI to run these in Concourse as a resource; on my infinite list of when-I-have-time-and-energy projects is to create tasks for the buildpack lifecycle binaries instead.

**Comment by hstenzel** [comment]
@jchesterpivotal I don't understand why we'd a build tool -- of any kind -- to be delivered as a resource instead of in a build task.

This is exactly the trouble that we got in with the Docker Image Resource. It conflated too many things. Consider that the role of the resource is to deal with the outside world. It gets things needed for the build. It puts the result in the correct location. It figures out when a new build needs to take place.

But when it actually does the build, that's when the trouble starts. And this happened with Docker, and would happen if cloud native buildpacks were implemented as a concourse resource.

I recently posted this one of the Concourse channels:

> 1) Use resources (get and put) for interacting with the outside world
> 2) Use tasks for introspective things and integration things. The task can inspect the gets and set the table for the puts; this broadly includes the "do it" step
> 3)  If 1 and 2 don't apply, see 1 and 2

And the trouble is that the Docker resource breaks the rules. It does a build but it does it in a push. I get the reasons for that, but some of the trouble I and others are having would be resolved if it were not so.

One way to solve the problem is to make a new class, a sort of build resource if you will. That's where the discussion started. Another way to solve the problem is to just embrace the current architecture and do the `docker build` (or `pack build`) in a task.

Now, there are other problems, but these boil down to it being a "bad thing" (tm) that it's difficult to create a OCI image inside an OCI container. There are emerging ways to solve this problem, and I don't think we have to be opinionated about how that is done. `img` is promising, especially once it no longer requires root. But dind, buildkit, buildah, and kaniko could all solve the problem in different ways.

Yet from the DX (Developer Experience) perspective, we still want to run the build in the same way in CI as a developer runs the build on their laptop. We just want to wrap it up in more reproducibility and proper versioning.

One of the other challenges is "How to deal with build systems that expect to fetch their own dependencies?" This is is something that crops up regularly for corporate users as we must use internal mirrors. It also violates the rules in that the build system is doing external communication. For instance, a Docker build must get its FROM images from the internal repo, then it must get the native packages it installs from the internal repo, then it must get the source level dependencies from the internal repo. I wonder if there's a good way to handle that . . .

**Comment by jchesterpivotal** [comment]
> @jchesterpivotal I don't understand why we'd a build tool -- of any kind -- to be delivered as a resource instead of in a build task.
>
> This is exactly the trouble that we got in with the Docker Image Resource. It conflated too many things. 

I agree and I gave this feedback to the folks working on the proof-of-concept. Hence my aside:

>  on my infinite list of when-I-have-time-and-energy projects is to create tasks for the buildpack lifecycle binaries instead.

On DX:

> Yet from the DX (Developer Experience) perspective, we still want to run the build in the same way in CI as a developer runs the build on their laptop. We just want to wrap it up in more reproducibility and proper versioning.

Yes! This is a key design motivation for Cloud Native Buildpacks. And it works exactly like that. The same source, fed to the same lifecycle and buildpack versions, will build exactly the same way either locally or remotely.

> One of the other challenges is "How to deal with build systems that expect to fetch their own dependencies?" This is is something that crops up regularly for corporate users as we must use internal mirrors. It also violates the rules in that the build system is doing external communication. 

Cloud Native Buildpacks are packaged into self-contained images called "builders", which encapsulate the lifecycle binaries, the buildpacks and dependencies that the buildpacks will provide. Motivated by the needs of disconnected environments, which has been a central requirement for the Cloud Foundry Buildpacks team since the beginning.
